Allow authors to remove option for unproctored exams

andy-armstrong opened the pull request:
## [TNL-5084](https://openedx.atlassian.net/browse/TNL-5084)
### Description

This story adds the ability for authors to specify that their courses do not support unproctored exams. This is useful so that the course author doesn't have to be concerned about unverified users logging in and finding out the exam questions, and they don't have to implement a separate exam for those users.

See the library's implementation PR too: https://github.com/edx/edx-proctoring/pull/315
### Screenshots

Here's the advanced setting in Studio:

![image](https://cloud.githubusercontent.com/assets/5985072/19865707/b91b49a0-9f73-11e6-95b9-22a5e0b89e5d.png)

Here's the resulting proctored course without the option to choose an unproctored exam:

![image](https://cloud.githubusercontent.com/assets/5985072/19736288/a889d12e-9b7c-11e6-98c5-7581411b7067.png)
### Sandbox
- [x] https://andy-armstrong.sandbox.edx.org/courses/course-v1:AndyA+AAPT+1/courseware/9b7aeac146d54fc7a552f2feb2b00f51/be3702e5245a4d6f96194449303bdbf3/

Note: use the verified@example.com user who is pre-verified for the course. The feature does not work with staff users who cannot take proctored exams.
### Testing
- [x] i18n
- [x] RTL
- [x] safecommit violation code review process
- [x] Unit, integration, acceptance tests as appropriate
### Reviewers

If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [ ] Code review: @dianakhuang 
- [ ] Code review: @robrap 
- [ ] Doc Review: @catong 
- [ ] UX review: @chris-mike 
- [x] Product review: @marcotuts 
### Post-review
- [ ] Rebase and squash commits

dianakhuang commented:
👍 from a code perspective. You will probably have to rebase off of master to fix the python issues.

robrap commented:
@andy-armstrong @catong: Based on the scrum discussion, a possible label might be "Allow proctoring opt-out".  No worries if you don't like it. :)

robrap commented:
jenkins run bokchoy

robrap commented:
jenkins run python

robrap commented on common/lib/xmodule/xmodule/seq_module.py:
```diff
@@ -130,6 +136,13 @@ def is_proctored_exam(self):
         return self.is_proctored_enabled
 
+    @property
+    def allow_unproctored_exam(self):
```
Reminder: Please update this function name and the comment below once @catong helps settle language questions, because it is confusing as is.

catong replied:
``allow_proctoring_optout```?

robrap commented on common/lib/xmodule/xmodule/course_module.py:
```diff
@@ -740,6 +740,16 @@ class CourseFields(object):
         scope=Scope.settings
     )
 
+    enable_unproctored_exams = Boolean(
```
Ultimately fix function name to match improved language.  It's too bad we don't have a good place for these settings.  I'll make sure that's on our list if it isn't already.

robrap commented on common/lib/xmodule/xmodule/seq_module.py:
```diff
@@ -117,6 +117,12 @@ class ProctoringFields(object):
+        """
+        Return course by course id.
+        """
+        return self.descriptor.runtime.modulestore.get_course(self.course_id)  # pylint: disable=no-member
```
Is this where we will soon be using course blocks api?  I imagine this is our only choice until the performance issue on that gets worked out.  Do we need a separate story around that, or do we wait until @nasthagiri's team gets to it?

robrap commented:
@andy-armstrong: Looks clean and simple enough.  I dropped a variety of questions that may not require any action.  I'd like to do another once renaming is complete and `unproctored` becomes something else.  Thanks.

andy-armstrong commented:
jenkins run python

catong commented on common/lib/xmodule/xmodule/course_module.py:
```diff
@@ -740,6 +740,16 @@ class CourseFields(object):
+    enable_unproctored_exams = Boolean(
+        display_name=_("Enable Unproctored Exams"),
+        help=_(
+            "Enter true or false. If this value is true, unproctored exams are enabled in your course. "
```
"Enter true or false. If this value is true, learners have the option to take proctored exams without proctoring."

catong commented on common/lib/xmodule/xmodule/course_module.py:
```diff
@@ -740,6 +740,16 @@ class CourseFields(object):
     )
 
+    enable_unproctored_exams = Boolean(
+        display_name=_("Enable Unproctored Exams"),
```
@andy-armstrong I like Rob's suggestion, tweaked it to:
"Allow Opting Out of Proctored Exams"

catong commented on common/lib/xmodule/xmodule/course_module.py:
```diff
@@ -740,6 +740,16 @@ class CourseFields(object):
+        display_name=_("Enable Unproctored Exams"),
+        help=_(
+            "Enter true or false. If this value is true, unproctored exams are enabled in your course. "
+            "If it is false, all exams will be required to be proctored."
```
@andy-armstrong per our conversation, can you please also modify the Help text for "Enable Timed Exams"? (Line 765). "Enter true or false. If this value is true, timed exams are enabled in your course. Regardless of this setting, timed exams are enabled if **Enable Proctored Exams** is set to true.

catong commented:
@andy-armstrong In the sandbox the verified user seems to be in a "verification pending" state, and I'm not able to see the various templates.

andy-armstrong commented:
@robrap @dianakhuang @catong @marcotuts thanks for all the great feedback. I've addressed your comments and have kicked-off a sandbox update. For the moment I've gone with @robrap's suggested name for the setting ("Allow Proctoring Opt-Out"), and I've updated the help text to match. We can do final tweaks to these strings on Monday. Please re-review when you have the time.

andy-armstrong commented:
Whoops, I missed @catong's comments. Sorry! I'll go with her recommendations.

catong commented on common/lib/xmodule/xmodule/course_module.py:
```diff
@@ -740,6 +740,18 @@ class CourseFields(object):
+    allow_proctoring_opt_out = Boolean(
+        display_name=_("Allow Opting Out of Proctored Exams"),
+        help=_(
+            "Enter true or false. If this value is true, the learner is given the option to choose "
```
@andy-armstrong Do you mind tweaking this help text a bit? 
"If this value is true, learners can choose to take proctored exams without proctoring. If this value is false, all learners must take the exam with proctoring. This setting only applies if proctored exams are enabled for the course."

robrap commented on lms/static/sass/course/courseware/_courseware.scss:
```diff
@@ -301,17 +301,11 @@ html.video-fullscreen {
-        }
-        &:last-child {
-          > .fa {
+          &.exam-without-proctoring {
```
Is this actually all exams that are not proctored, timed and otherwise?  Are these timed exams?  Are these proctored exams where the learner opted out of proctoring?
If it is the latter, maybe?:
```
 &.exam-with-proctoring-optout {
```
Also, can you help me understand how this is used without a reference somewhere else?  Thanks.

andy-armstrong replied:
The way this works is that the edx-proctoring repo renders a button with this class, and then this Sass rule in edx-platform changes the way it renders to not show as a blue button. The old code always showed the last button as gray, but that doesn't work when we have only a single button.

I think I'll rename it as ``secondary-action`` as the only meaning is that this is not the primary button.

andy-armstrong replied:
In fact, I think I'll switch the logic and have a ``primary-action`` class instead...

robrap replied:
Thanks.  I forgot about the other PR. :)  👍  with this change.  It's too bad this can't live with its reference.

andy-armstrong commented:
@robrap @dianakhuang @catong @marcotuts I've addressed all the feedback, squashed the commits, and updated my sandbox. This should be ready for final review now.

dianakhuang commented:
👍  from me.

catong commented:
Looks great, @andy-armstrong 👍 from me

andy-armstrong commented:
jenkins run python

marcotuts commented:
👍
